lib/index.js

```diff
@@ -40,18 +38,33 @@ class Parser extends EventEmitter {
       this.emit('nmea0183', sentence.trim())
     }
 
+    try {
+      var result = parseSentence(this, sentence, this.options)
+      if (typeof result === 'object' && result !== null) {
+        this.emit('signalk:delta', result || {})
+      }
+      // Return value kept for backwards compatibility. The usage of
+      // parseImmediate("") is recommended instead.
+      return Promise.resolve({ delta: result })
+    }
+    catch(e) {
+      // FIXME: Is this really useful? Maybe should just throw the exception
```

Comment:
To be backwards compatible this should return a rejected Promise?

Reply:
Returning null is sufficient here I think - are we throwing errors somewhere?

Reply:
> To be backwards compatible this should return a rejected Promise?
> Returning null is sufficient here I think - are we throwing errors somewhere?

I saw that tonight too while trying to remove some changes and simplify this PR. It's more complicated than I thought to be really backwards compatible.

As far as I can tell, current code throws exception in a few places that are converted in `emit('error')` [in `lib/index.js`](https://github.com/SignalK/signalk-parser-nmea0183/blob/master/lib/index.js#L53):

```
  parser
    .parse(data)
    .then(result => {
      if (result != null) {
        console.log(JSON.stringify(result.delta))
      }
    })
    .catch(e => {
      console.error('Encountered an error:', e.message)
    })
```

In `bin/nmea0183-signalk` we have:
```
    return parseSentence(this, sentence, this.options)
      .then(result => {
        if (typeof result === 'object' && result !== null) {
          // this.emit('signalk:full', result.full || {})
          this.emit('signalk:delta', result.delta || {})
        }
        return result
      })
      .catch(e => {
        this.emit('error', e)
      })
```

The `promise.catch` is never reached because we already have a catch in `lib/index.js`. Instead, the `promise.then()` of `bin/nmea0183-signalk` is called with whatever was returned from the `catch` statement, in this case we have no return statement so the catch returns `undefined`.

So to be backwards compatible, we need to:
 - `emit(error)`
 - Return a `Promise.resolve(undefined)`

Let me know if you disagree. I will update the code to match this.

Reply:
Well, the cli utility is internal to this module, so we don't need to pay attention to being backwards compatible there.

Being backwards compatible is (was?) more about being possibly more confident that this not cause regression errors than supporting a multitude of projects that rely on current functionality.

I sincerely doubt anything but your project and sk server use this to any meaningful extent. We can break backwards compatibility and adjust the server's code accordingly if we want to. In fact to get real benefit from the synchronous api to fix the timestamp related problem I mentioned earlier we should adjust the server's code.

So let's make the new api good with less convoluted error handling (maybe a test or two for that) and not worry too much.